### PR TITLE
Dev hwcomposer3

### PR DESCRIPTION
**Pull request by manxiaoliang**
Upgrade the graphics composer HAL to AIDL version in android 13

### CODE REVIEW

File: hwc3-service/default/Android.bp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        "-Werror",
+        "-DLOG_TAG=\"hwc3\"",
+	],
+    cppflags: [
+        "-Wextra",
+        "-Wunused",
+        "-Wunreachable-code",
+        "-Wno-sign-compare",
+    ],
+	local_include_dirs:[
```
Comment (feijiang1) [inline hwc3-service/default/Android.bp:41]: coding style issue, need align correctly

File: hwc3-service/default/Android.mk [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	service.cpp
+
+ifeq ($(BOARD_USES_HWC_SERVICES),true)
+LOCAL_CFLAGS += -DUSES_HWC_SERVICES
+LOCAL_SHARED_LIBRARIES += libExynosHWCService
+LOCAL_HEADER_LIBRARIES += libbinder_headers
+endif
+
+LOCAL_VINTF_FRAGMENTS = hwc3-default.xml
+LOCAL_INIT_RC := hwc3-pixel.rc
```
Comment (feijiang1) [inline hwc3-service/default/Android.mk]: why pixel is named?

File: hwc3-service/default/Android.bp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+//     // to get the below license kinds:
+//     //   SPDX-license-identifier-Apache-2.0
+//     default_applicable_licenses: ["hardware_interfaces_license"],    
+// }
+
+cc_binary {
+    name: "android.hardware.graphics.composer3-service.intel",
+	relative_install_path: "hw",
+    init_rc: ["android.hardware.graphics.composer3-service.intel.rc"],
+    vintf_fragments: ["android.hardware.graphics.composer3-service.intel.xml"],
```
Comment (feijiang1) [inline hwc3-service/default/Android.bp:28]: remove xml file

File: hwc3-service/default/Android.bp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    ],
+	local_include_dirs:[
+		"include",
+        "impl",
+	],
+    include_dirs: [ 
+    ],
+
+    header_libs: [
+        "android.hardware.graphics.composer@2.1-passthrough",
```
Comment (feijiang1) [inline hwc3-service/default/Android.bp:49]: double check it

File: hwc3-service/default/android.hardware.graphics.composer3-service.intel.xml [hunk 5]
```diff
@@ -6,5 +6,6 @@
             <name>IComposer</name>
```
Comment (feijiang1) [inline hwc3-service/default/android.hardware.graphics.composer3-service.intel.xml:6]: remove it

File: hwc3-service/default/impl/HalImpl.cpp [hunk 6]
```diff
@@ -101,27 +96,34 @@ void seamlessPossible(hwc2_callback_data_t callbackData, hwc2_display_t hwcDispl
...
+    if (mDevice) {
+        mDevice->common.close(&mDevice->common);
+        mDevice = nullptr;
+    }
 }
 
 void HalImpl::initCaps() {
     uint32_t count = 0;
-    mDevice->getCapabilities(&count, nullptr);
+    mDevice->getCapabilities(mDevice,&count, nullptr);
```
Comment (feijiang1) [inline hwc3-service/default/impl/HalImpl.cpp:123]: double check, if need, add space

File: hwc3-service/default/impl/HalImpl.cpp [hunk 7]
```diff
@@ -168,236 +268,246 @@ void HalImpl::getCapabilities(std::vector<Capability>* caps) {
...
+                               reinterpret_cast<hwc2_function_pointer_t>(hook::hotplug));
+    mDispatch.registerCallback(mDevice, HWC2_CALLBACK_REFRESH, this,
+                               reinterpret_cast<hwc2_function_pointer_t>(hook::refresh));
+    mDispatch.registerCallback(mDevice, HWC2_CALLBACK_VSYNC_2_4, this,
+                               reinterpret_cast<hwc2_function_pointer_t>(hook::vsync));
+    mDispatch.registerCallback(mDevice, HWC2_CALLBACK_VSYNC_PERIOD_TIMING_CHANGED, this,
+                               reinterpret_cast<hwc2_function_pointer_t>(hook::vsyncPeriodTimingChanged));
+    mDispatch.registerCallback(mDevice, HWC2_CALLBACK_SEAMLESS_POSSIBLE, this,
+                               reinterpret_cast<hwc2_function_pointer_t>(hook::seamlessPossible));
 }
```
Comment (feijiang1) [inline hwc3-service/default/impl/HalImpl.cpp:291]: add ToDo

File: hwc3-service/default/impl/HalImpl.cpp [hunk 8]
```diff
@@ -168,236 +268,246 @@ void HalImpl::getCapabilities(std::vector<Capability>* caps) {
...
-
-    // unregister HWC3 Callback
-    mDevice->registerHwc3Callback(IComposerCallback::TRANSACTION_onVsyncIdle, this, nullptr);
-
+    mDispatch.registerCallback(mDevice, HWC2_CALLBACK_HOTPLUG, this, nullptr);
+    mDispatch.registerCallback(mDevice, HWC2_CALLBACK_REFRESH, this, nullptr);
+    mDispatch.registerCallback(mDevice, HWC2_CALLBACK_VSYNC_2_4, this, nullptr);
+    mDispatch.registerCallback(mDevice, HWC2_CALLBACK_VSYNC_PERIOD_TIMING_CHANGED, this, nullptr);
+    mDispatch.registerCallback(mDevice, HWC2_CALLBACK_SEAMLESS_POSSIBLE, this, nullptr);
     mEventCallback = nullptr;
```
Comment (feijiang1) [inline hwc3-service/default/impl/HalImpl.cpp:299]: add ToDo for TRANSACTION_onVsyncIdle

File: hwc3-service/default/impl/TranslateHwcAidl.h [hunk 9]
```diff
@@ -64,31 +62,31 @@ inline void translate(const hwc_client_target_property& in, ClientTargetProperty
...
+//     case HwcMountOrientation::ROT_270:
+//         out = common::Transform::ROT_270;
+//         break;
+
+//     default:
+//         LOG(WARNING) << "unrecoganized display orientation: " << static_cast<uint32_t>(in);
+//         out = common::Transform::NONE;
+//         break;
+//     }
+// }
```
Comment (feijiang1) [inline hwc3-service/default/impl/TranslateHwcAidl.h:89]: remove them

**Review by feijiang1 — COMMENTED** [review]
some comments

File: hwc3-service/default/impl/HalImpl.cpp (hunk 6 above)
Comment (manxiaoliang) [reply hwc3-service/default/impl/HalImpl.cpp:123]: vendor/intel/external/drm-hwcomposer/.ci/android_headers/hardware/hwcomposer2.h
typedef struct hwc2_device {
    ......
    void (*getCapabilities)(struct hwc2_device* device, uint32_t* outCount,
            int32_t* /*hwc2_capability_t*/ outCapabilities);
    ......
} hwc2_device_t;

File: hwc3-service/default/Android.bp (hunk 4 above)
Comment (manxiaoliang) [reply hwc3-service/default/Android.bp:49]: don't need android.hardware.graphics.composer@2.1-passthrough

**Pull request closed without merging**